whylog/log_reader/__init__.py

```diff
@@ -1,21 +1,81 @@
+import itertools
 from abc import ABCMeta, abstractmethod
+from collections import defaultdict
 
 import six
 
+from whylog.front import FrontInput
+from whylog.log_reader.exceptions import NoLogTypeError
+from whylog.log_reader.investiagtion_utils import InvestigationUtils
+from whylog.log_reader.searchers import BacktrackSearcher
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractLogReader(object):
     @abstractmethod
     def get_causes(self, front_input):
         pass
 
+    @abstractmethod
+    def get_causes_tree(self, front_input):
+        pass
+
 
 class LogReader(AbstractLogReader):
     def __init__(self, config):
         self.config = config
 
     def get_causes(self, front_input):
-        pass
+        input_log_type = self.config.get_log_type(front_input)
+        if not input_log_type:
+            raise NoLogTypeError(front_input)
+        investigation_plan = self.config.create_investigation_plan(front_input, input_log_type)
+        manager = SearchManager(investigation_plan)
+        return manager.investigate()
 
     def get_causes_tree(self, front_input):
         pass
+
+
+class SearchManager(object):
+    def __init__(self, investigation_plan):
+        self._investigation_plan = investigation_plan
+
+    @classmethod
+    def _save_clues_in_normal_dict(cls, collector):
```

Comment:
Not exactly; the resulting dictionary does not contain itertools.chain's, but lists of clues instead. It is because itertools.chain is a generator, so it is impossible to use it more then once. At the moment I cannot ensure that it will not be used twire or more times... maybe yes, but using list is less risky.